### PR TITLE
sync connect flows

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/488551667048375/1204039315134720/f 

### Description
Implements 2 connect flows:
- one authenticated device connecting one unauthenticated device without camera.
- 2 unauthenticated devices joining together on a sync account

### Steps to test this PR
For all tests you will need 2 devices here.

_2 unauthenticated devices_
- [x] Install the app on 2 devices
- [x] on device A: go to settings -> sync -> Show Connect QR code
- [x] it will start polling data every 7 seconds
- [x] It will show a toast with the error until we perform connect on the second device
- [x] on device B: got to settings -> sync
- [x] press read Connect QR code
- [x] read QR from device A
- [x] It will create an account
- [x] device B should join the account soon
- [x] ensure userId and primary key matches

_1 unauthenticated device_
- [x] Install the app on 2 devices
- [x] on device A: go to settings -> sync -> Show Connect QR code
- [x] it will start polling data every 7 seconds
- [x] It will show a toast with the error until we perform connect on the second device
- [x] on device B: got to settings -> sync
- [x] create an account
- [x] press read Connect QR code
- [x] read QR from device A
- [x] device B should join the account soon
- [x] ensure userId and primary key matches

 
### UI changes
| Before  | After |
| ------ | ----- |
!(Upload before screenshot)|(Upload after screenshot)|

### CODE REVIEW

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt [hunk 1]
```diff
@@ -142,6 +146,87 @@ class AppSyncRepository @Inject constructor(
...
+            )
+
+        return when (result) {
+            is Result.Error -> {
+                result
+            }
+            is Result.Success -> {
+                val decryptResult = nativeLib.decrypt(result.data.protected_encryption_key, preLogin.stretchedPrimaryKey)
+                if (decryptResult.result != 0L) return Result.Error(code = decryptResult.result.toInt(), reason = "Decrypt failed")
+                Timber.i("SYNC decrypt: decoded secret Key: ${decryptResult.decryptedData}")
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt:218]: wanna get rid of this?

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncService.kt [hunk 2]
```diff
@@ -67,6 +80,15 @@ data class Logout(
     val device_id: String,
 )
 
+data class ConnectKey(
+    val encrypted_recovery_key: String,
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncService.kt:84]: ditto about the JSON field name

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncService.kt [hunk 3]
```diff
@@ -67,6 +80,15 @@ data class Logout(
...
+)
+
+data class Connect(
+    val device_id: String,
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncService.kt:88]: ditto about the JSON field name

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncService.kt [hunk 4]
```diff
@@ -67,6 +80,15 @@ data class Logout(
...
+    val encrypted_recovery_key: String,
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncService.kt:89]: ditto about the JSON field name

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupViewModel.kt [hunk 5]
```diff
@@ -156,4 +160,49 @@ constructor(
...
+    }
+
+    fun onConnectStart() {
+        viewModelScope.launch(dispatchers.io()) {
+            val qrCode = syncRepository.getConnectQR()
+            updateViewState()
+            command.send(ShowQR(qrCode))
+            var polling = true
+            while (polling) {
+                delay(7000)
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupViewModel.kt:186]: let's not forget to clean this up later

**Review by malmstein — APPROVED** [review]
works as expected, just a couple nits

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupViewModel.kt (hunk 5 above)
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/SyncInitialSetupViewModel.kt:186]: the screen will be removed, so its fine

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt [hunk 6]
```diff
@@ -142,6 +146,87 @@ class AppSyncRepository @Inject constructor(
         return Adapters.recoveryCodeAdapter.toJson(LinkCode(RecoveryCode(primaryKey, userID)))
     }
 
+    override fun getConnectQR(): String {
+        val prepareForConnect = nativeLib.prepareForConnect()
+        val deviceId = syncDeviceIds.deviceId()
+        syncStore.deviceId = deviceId
+        syncStore.primaryKey = prepareForConnect.publicKey
+        syncStore.secretKey = prepareForConnect.secretKey
+        return Adapters.recoveryCodeAdapter.toJson(LinkCode(connect = ConnectCode(device_id = deviceId, secret_key = prepareForConnect.publicKey)))
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncRepository.kt]: No strong opinion, I don't feel strong about not having it here, but I don't see a reason to have `Result` across all return values.

as mentioned, I plan to make a review later of the syncrepository interface, and I have captured this question in https://app.asana.com/0/0/1204139091554738/f

**Comment by cmonfortep** [comment]
CI checks failing due to TEST_QUOTA_EXCEEDED when running tests. Gonna merge the PR into the next branch, and wait to merge in develop until we fix it. 

If any issue is detected, I will address it there.

**Pull request merged**